File: 이준영/백준/Gold/17836. 공주님을 구해라！/공주님을 구해라！.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Stack;
import java.util.StringTokenizer;

public class Main {
	public static void main(String[] args) throws Exception{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		int[] dx = {0,0,1,-1};
		int[] dy = {1,-1,0,0};
		int N = Integer.parseInt(st.nextToken());
		int M = Integer.parseInt(st.nextToken());
		int T = Integer.parseInt(st.nextToken());
		int minimum=T+1;
		int[][] arr = new int[N][M];
		Stack<int[]> stack = new Stack<>();
		boolean[][] visited = new boolean[N][M];
		for(int i=0;i<N;i++) {
			st = new StringTokenizer(br.readLine());
			for(int j=0;j<M;j++) {
				arr[i][j]=Integer.parseInt(st.nextToken());
				
			}
		}
		int time=1;
		int[] v = {0,0};
		stack.push(v);
		visited[0][0]=true;
		outer:while(time<=T) {
			Stack<int[]> stack2 = new Stack<>();
			while(!stack.isEmpty()) {
				int[] tmp=stack.pop();
				for(int i=0;i<4;i++) {
					int ny = tmp[0]+dy[i];
					int nx = tmp[1]+dx[i];
					if(0<=ny&&ny<N&&0<=nx&&nx<M&&!visited[ny][nx]&&arr[ny][nx]!=1) {
						visited[ny][nx]=true;
						if(ny==N-1&&nx==M-1) {
							minimum = Math.min(minimum,time);
							break outer;
						} else if(arr[ny][nx]==2) {
							minimum=Math.min(minimum, time+(N+M-2-ny-nx));
							int[] tmp2 = {ny,nx};
							stack2.push(tmp2);
						} else {
							int[] tmp2 = {ny,nx};
							stack2.push(tmp2);
						}
					}
				}
			}
			stack=stack2;
			time+=1;
		}
		if(minimum==T+1) {
			System.out.println("Fail");
		} else {
			System.out.println(minimum);
		}
		
	}
}

File: 이준영/백준/Gold/17836. 공주님을 구해라！/공주님을 구해라！.py
import sys
from collections import deque

input = sys.stdin.readline

N, M, T = map(int, input().split())
arr = [list(map(int, input().split())) for _ in range(N)]
visited = [[0] * M for _ in range(N)]

time = T + 1
visited[0][0] = 1
dq = deque([(0, 0, 1)])
dy = [0, 0, 1, -1]
dx = [1, -1, 0, 0]

while dq:
    y, x, t = dq.popleft()
    if t >= time:
        break

    for k in range(4):
        ny = y + dy[k]
        nx = x + dx[k]
        if 0 <= ny < N and 0 <= nx < M and arr[ny][nx] != 1 and visited[ny][nx] == 0:
            if ny == N - 1 and nx == M - 1:
                time = min(time, t)
                break
            visited[ny][nx] = 1
            if arr[ny][nx] == 2:
                if t + abs(ny - N + 1) + abs(nx - M + 1) <= T:
                    time = min(t + abs(ny - N + 1) + abs(nx -M + 1), time)
            else:
                dq.append((ny, nx, t + 1))

if time == T + 1:
    print("Fail")

else:
    print(time)
